File: java/ankit_kanyal/devnest_problems/phase_1/InfixToPostFix.java
static int precedence(char c){

switch (c){
    case '+':
    case '-':
        return 1;
    case '*':
    case '/':
        return 2;
    case '^':
        return 3;

}

return -1;
}

static String solve(String infix){

//CODE HERE

String result = "";

Stack<Character> stack = new Stack<>();

for (int i = 0; i <infix.length() ; i++) {
    char c = infix.charAt(i);

//check if char is operator
    if(precedence(c)>0){
        while(stack.isEmpty()==false && precedence(stack.peek())>=precedence(c)){
            result += stack.pop();
        }
        stack.push(c);

    }else if(c==')'){
        char x = stack.pop();
        while(x!='('){
            result += x;
            x = stack.pop();
        }
    }else if(c=='('){
        stack.push(c);

    }else{
    //character is neither operator nor ( 
        result += c;
    }
}

for (int i = 0; i <=stack.size() ; i++) {
    result += stack.pop();

}

return result;
}

File: python/vedant_bhatnagar/DevsnestProblems_THA/Phase 1/infixtopostfix.py
def solve(infix):
    # CODE HERE
    postfix = ""
    op = {"+":1, "-":1, "*":2, "/":2, "^":3}
    st = []
    for i in infix:
        # print(i)
        if i not in op and i != "(" and i != ")":
            postfix += i
            continue
        if i == "(":
            st.append(i)
            continue
        if i == ")":
            # print(st)
            while st[-1] != "(":
                el = st.pop()
                postfix+=el
            st.pop()
            continue
        if i in op:
            while st:
                st_top = st[-1]
                if st_top == "(":
                    st.append(i)
                    break
                else:
                    if op[st_top] >= op[i]:
                        postfix += st_top
                        st.pop()
                    else:
                        st.append(i)
                        break
            if len(st) == 0:
                st.append(i)
    while st:
        postfix += st.pop()
    return postfix